src/legacy/core_plugins/embeddable_api/public/panel/panel_header/panel_actions/inspect_panel_action.tsx

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { EuiIcon } from '@elastic/eui';
+import React from 'react';
+
+import { i18n } from '@kbn/i18n';
+import { Inspector } from 'ui/inspector';
+import { Action, ActionContext } from '../../../actions';
+
+export const INSPECT_PANEL_ACTION_ID = 'openInspector';
+
+export class InspectPanelAction extends Action {
+  public readonly type = INSPECT_PANEL_ACTION_ID;
+  constructor() {
+    super(INSPECT_PANEL_ACTION_ID);
+    this.order = 10;
+  }
+
+  public getDisplayName() {
+    return i18n.translate('kbn.embeddable.panel.inspectPanel.displayName', {
+      defaultMessage: 'Inspect',
+    });
+  }
+
+  public getIcon() {
+    return <EuiIcon type="inspect" />;
+  }
+
+  public async isCompatible({ embeddable }: ActionContext) {
+    return Inspector.isAvailable(embeddable.getInspectorAdapters());
+  }
+
+  public async execute({ embeddable }: ActionContext) {
+    const adapters = embeddable.getInspectorAdapters();
+
+    if (!(await this.isCompatible({ embeddable })) || adapters === undefined) {
```

Comment:
As I see it, there are two stages to an embeddable+action life cycle.

1. Setup - this is the stage where we choose all of the actions to be associated with the embeddable. We choose actions according to viewMode, user settings, etc. At this stage we only have an **Embeddable**.
2. Execution - this is where the user actually interacts with the embeddable and therefore we have the **TriggerContext**. At this stage, it's just about handling input and calling the correct stuff.

If my understanding is correct, why would an action be executed and then be incompatible?
If the behavior in this code would happen, the user would try to execute an action we offered them, and then get an "action incompatible" error. So why would we show it to them in the first place?

P.S. I see this behavior in other actions as well - so this question applies to all of them.

Reply:
You do actually have the `triggerContext` at the time of setup as well as execution.  User clicks something which emits the trigger with dynamic trigger context, mapped actions gathered, and if more than one are mapped to that trigger, the context menu is created and shown.  The confusion I think is that anything attached to `SHOW_CONTEXT_MENU` should not rely on any triggerContext - it will never get any, not during setup, nor execution.

But, this will be a question we'll have to ask for phase 2.  I just want to be careful making too many assumptions and changes about phase 2 before we actually start implementing it.  

> If my understanding is correct, why would an action be executed and then be incompatible?

it shouldn't ever be, but a developer could bypass the isCompatible check, so this is just a little extra protection, which is actually helpful for typescript anyway.